File: OpenGL/Windows/ProgramablePipeline/24-SolarSystem/src/MyStack.cpp
#pragma once
#include <Windows.h>
#include "MyStack.h"
#include "OGL.h"

struct CTM* pTop = NULL;

void pushMatrix(mat4 data) {

	CTM* temp = NULL;

	temp = (CTM*)malloc(sizeof(CTM));
	if (!temp) {
		fprintf(gpFile, "\nERROR IN STACK FAILED TO ALLOCATE MEMORY\n");
		DestroyWindow(ghwnd);
	}

	temp->data = data;
	temp->link = pTop;
	pTop = temp;
}


void popMatrix(mat4 *data) {

	CTM* temp = NULL;

	if (pTop == NULL) {
		fprintf(gpFile, "\nERROR IN STACK FAILED TO ALLOCATE MEMORY\n");
		DestroyWindow(ghwnd);
	}

	temp = pTop;
	*data = temp->data;
	pTop = pTop->link;
	temp->link = NULL;
	free(temp);
	temp = NULL;
}


void deleteStack() {
	
	CTM* temp = NULL;


	while (temp != NULL) {
		temp = pTop;
		
		pTop = pTop->link;
		temp->link = NULL;

		free(temp);
		temp = NULL;
	}

	free(pTop);
	pTop = NULL;
}

File: Android-EGL/34-SolarSystem/src/main/java/com/example/ogl27/MyStack.java
package com.example.ogl27;

class MyStack {
   
    private float[][] stack;
    private int top;
    private int size; 

    private final int FAIL = -1;
    private final int SUCCESS = 0;

    public MyStack(int size) {
        this.size = size;
        this.top = -1;
        stack = new float[this.size][16];
    }

    public int pushMatrix(float[] arr) {
        
        if(top == size-1) {
            return FAIL;
        }
        top++;
        for(byte i = 0; i < 16; i++) {
            stack[top][i] = arr[i];
        }
        return SUCCESS;
    }

    public int popMatrix(float[] arr) {

        if(top == -1) {
            return FAIL;
        }

        for(byte i = 0; i < 16; i++) {
            arr[i] = stack[top][i];
        }
        top--;
        return SUCCESS;
    }
}